tests: drivers: counter: Fix test for nrf51

nordic-krch opened the pull request:
Fixes test which did not take into account that counter may
wrap (nrf51 has 16bit timers) and was setting alarm to wrong
value. Error was not seen on platforms with higher top value.

Fixes #24369.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: drivers: counter: Fix test for nrf51

Fixes test which did not take into account that counter may
wrap (nrf51 has 16bit timers) and was setting alarm to wrong
value. Error was not seen on platforms with higher top value.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented on tests/drivers/counter/counter_basic_api/src/test_counter.c:
```diff
@@ -797,7 +799,8 @@ static void test_cancelled_alarm_does_not_expire_instance(const char *dev_name)
 
 		k_busy_wait(us/2 + i);
 
-		alarm_cfg.ticks = alarm_cfg.ticks + 2*alarm_cfg.ticks;
```
test was suppose to set some alarm in the future (it is cancelled anyway) and for non wrapping (within test) counters this formula was ok but it was actually wrong and initial intention was that it is `2*ticks` (2 ms from now).

anangl replied:
> and initial intention was that it is `2*ticks` (2 ms from now).

The commit message would be a better place for such additional explanation.

mbolivar-nordic commented:
The test is still failing for me. https://gist.github.com/mbolivar-nordic/8410c20dda61345b67e806d340aa6c0c

mbolivar-nordic reviewed (changes requested):
-1 since this is not a fix in my testing

nordic-krch committed:
tests: drivers: counter: Fix validation of the alarm value

Test is checking if alarm handler is executed at request time or later.
However, validation did not take into account wrapping of the counter.
Fixed by taking into account case where counter wraps.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
>-1 since this is not a fix in my testing

That test was spurious as it was failing only when alarm expired when counter wrapped. That was revealing a bug in the test - validation that alarm came not earlier than requested. Fixed.

mbolivar-nordic commented on tests/drivers/counter/counter_basic_api/src/test_counter.c:
```diff
@@ -254,23 +254,33 @@ void test_set_top_value_without_alarm(void)
+	/* Arbitrary limit for alarm processing - time between hw expiration
+	 * and read-out from counter in the handler.
+	 */
+	static const u64_t processing_limit_us = 1000;
```
Why static? Just curious.

nordic-krch replied:
puts variable in flash and not on stack.

mbolivar-nordic replied:
Are we really so close to blowing the stack on this simple test case? I'm not objecting to it, it just looked weird.

mbolivar-nordic commented on tests/drivers/counter/counter_basic_api/src/test_counter.c:
```diff
@@ -254,23 +254,33 @@ void test_set_top_value_without_alarm(void)
 
+	zassert_true(diff < counter_us_to_ticks(dev, processing_limit_us),
+			"Unexpected distance between reported alarm value(%u) "
+			"and actual counter value (%u), top:%d",
```
Would be good to include processing_limit_us in the output here too, I think.

mbolivar-nordic reviewed (approved):
Test is passing now so I'm happy, thanks! A couple of non-blocking style things.

nordic-krch commented:
i'm getting counter test failures on `mps2_an385`. However, locally those tests are passing. @nashif do you know who maintains counter for `mps_an385`?

carlescufi commented:
@nordic-krch there seems to be a regression in this PR: https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/69627/3/console

I have re-run CI but the failure re-appeared so this is not a false positive.

nordic-krch commented:
@carlescufi yes, on CI is it consistent but on my setup it is not. This test is passing on my machine and changes do not touch failing test so i think it is also failing on master.

nordic-krch commented:
created bug report for failing test: #24635

nordic-krch commented:
@carlescufi can you merge? We accidentally got green build. Root cause is known. #22904 should eventually fix it.
